### PR TITLE
[EPIC] Schabowski: Fixes for bugs revealed by Coinlist, EthCapeTown, and EthNY hackathons

### LINKED ISSUE #833: Bob cannot retrieve the same data twice

### LINKED ISSUE #920: Some controller endpoints fail when trying to serialize output.
This is an insidious issue.  

We take string labels for our endpoints (as JSON).  These strings get encoded as b64.  Then, when the response is parsed, they are rendered back to UTF-8 for display as JSON.

This means that 1 out of every 8 labels is invalid.  For example, `'516d593559505355376d454b61374751577146467a47754658396d516a685674716b7663744b376b4b666a35336d'` is not valid, because its b64 representation can't be serialized as UTF-8.

Here's a discussion of this matter: https://ptb.discordapp.com/channels/411401661714792449/411401661714792451/564353305887637517

@KPrasch suggests in #508 that we adopt UTF-8; I can now fully concur.

### LINKED ISSUE #925: Implement epic-v1
We've talked about our new plan for epic branches.  OK, I'll start.

Here's what I intend to do:

* Open a new epic branch, with an opening commit (the **opener**).
* Open a Pull Request from this branch against master.
* Add a checklist in the description, describing (albeit not necessarily committing) to the conditions which will warrant merge.

Rules I think make sense:
* I'll make up a silly title for myself (I'm going with "chaperone"), and I'll put, in the description, "chaperones".  Other people may add themselves.  Doing so is a commitment to do the following things:
** Conduct rebases over master. 
** Conduct merges into the epic branch.
** Work with any contemporaneous epickers to manage which PRs are in and which are out.
* *All merge commits* within this epic must include the **opener** - in other words, no merge commits are allowed between sub-PRs; only from sub-PRs to the epic.  *All other commits* must be rebased such that this commit is in the history, with a lineal sequence maintained by the author until merge.
* All reviews for PRs going in are subject to some kind of checklist or templating.  It is apparently not possible to auto-populate the review field on github as it is with the PR description field.  See: https://github.community/t5/How-to-use-Git-and-GitHub/Pull-Request-Review-Template/td-p/6857

### LINKED ISSUE #941: Finnegan's Wake demo... only works with --debug.
Something has changed at some point.  Now, the Finnegan's Wake demo appears to only work with the 3-ursulas docker container if the Ursulas contained therein are launched with `--debug`.  I mean... what?

For example: Why does 638a4df78957874141a8178e9bac2fac071ca796 work but fc16ec923e7bb2c62da78c167a38b789bfa06f4f does not?

### PR DESCRIPTION
**Pull request by jMyles**
* Chaperones: @jMyles 

- [x] Establishes new epic process (Fixes #925)
- [x] Documents epics in pre-release period
- [x] Corrects Character Control encoding (Fixes #920)
- [x] Fixes Bob's multiple retrieval issues (Fixes #833)
- [x] Corrects side-effects of `--debug` (Fixes #941)

---
*Named for **Günter Schabowski**, the somewhat hapless spokesperson for the Sozialistische Einheitspartei Deutschlands who announced on 9 Nov 1989 that travel restrictions from East to West Berlin had been lifted effective immediately - he was supposed to announce that they were going to be slowly lifted over several months.  His announcement led to a surge of many thousands of people to the wall - many with hammers and chisels - and, a few hours later, into West Berlin for the first time in decades.*

*Runner-up name: Ginger, based on what @jMyles was drinking when he opened it.*

### CODE REVIEW

**Comment by cygnusv** [comment]
Hey @jMyles, what about adding my fix to #245 here?
Here's the commit: a7d9d7cbbe963e384f993e18c6f1f8c4ec5772a8

**Comment by KPrasch** [comment]
I think #941 Will also be fixed in a parallel effort (#951)

**Comment by cygnusv** [comment]
@jMyles Just bumping this:

> Hey @jMyles, what about adding my fix to #245 here?
> Here's the commit: [a7d9d7c](https://github.com/nucypher/nucypher/commit/a7d9d7cbbe963e384f993e18c6f1f8c4ec5772a8)

File: tests/characters/test_bob_joins_policy_and_retrieves.py [hunk 1]
```diff
@@ -99,23 +98,45 @@ def test_bob_joins_policy_and_retrieves(federated_alice,
 
     assert plaintext == delivered_cleartexts[0]
 
-    # FIXME: Bob tries to retrieve again
```
Comment (KPrasch) [inline tests/characters/test_bob_joins_policy_and_retrieves.py]: :+1:

File: tests/cli/test_mixed_configurations.py [hunk 2]
```diff
@@ -141,7 +141,7 @@ def test_coexisting_configurations(click_runner,
     #
 
     # Run an Ursula amidst the other configuration files
-    run_args = ('ursula', 'run', '--dry-run',
+    run_args = ('ursula', 'run', '--dry-run', '--interactive',
```
Comment (KPrasch) [inline tests/cli/test_mixed_configurations.py]: :+1:

File: tests/fixtures.py [hunk 3]
```diff
@@ -284,9 +284,19 @@ def enacted_blockchain_policy(idle_blockchain_policy, blockchain_ursulas):
 
 @pytest.fixture(scope="module")
 def capsule_side_channel(enacted_federated_policy):
-    enrico = Enrico(policy_encrypting_key=enacted_federated_policy.public_key)
-    message_kit, _signature = enrico.encrypt_message(b"Welcome to the flippering.")
-    return message_kit, enrico
+    class _CapsuleSideChannel:
```
Comment (KPrasch) [inline tests/fixtures.py]: Another Side Channel... We are accumulating them!

File: nucypher/network/server.py [hunk 4]
```diff
@@ -181,10 +182,13 @@ def learn_about_announced_nodes():
...
                     this_node.suspicious_activities_witnessed['vladimirs'].append(node)
-
+                except NodeSeemsToBeDown as e:
+                    # This is a rather odd situation - this node *just* contacted us and asked to be verified.  Where'd it go?  Maybe a NAT problem?
+                    log.info(f"Node announced itself to us just now, but seems to be down: {node}.  Response was {e}.")
+                    log.debug(f"Phantom node certificate: {node.certificate}")
                 # Async Sentinel
                 except Exception as e:
-                    log.critical(str(e))
+                    log.critical(f"This exception really needs to be handled differently: {e}")
```
Comment (KPrasch) [inline nucypher/network/server.py]: OK..... haha - Perhaps an Issue #?

File: tests/characters/control/test_web_control.py [hunk 5]
```diff
@@ -4,12 +4,40 @@
...
+click_runner = CliRunner()
+
+
+@pytest.mark.skip("920")
+def test_label_whose_b64_representation_is_invalid_utf8():
+    # In our Discord, user robin#2324 (github username @robin-thomas) reported certain labels
+    # break Bob's retrieve endpoint.
+    # convo starts here: https://ptb.discordapp.com/channels/411401661714792449/411401661714792451/564353305887637517
+
+    bad_label = '516d593559505355376d454b61374751577146467a47754658396d516a685674716b7663744b376b4b666a35336d'
```
Comment (tuxxy) [inline tests/characters/control/test_web_control.py]: :+1:

File: tests/characters/control/test_web_control.py [hunk 6]
```diff
@@ -4,12 +4,40 @@
...
+from nucypher.characters.control.serializers import AliceControlJSONSerializer
 from nucypher.crypto.kits import UmbralMessageKit
 from nucypher.crypto.powers import DecryptingPower
 from nucypher.policy.models import TreasureMap
 
+click_runner = CliRunner()
+
+
+@pytest.mark.skip("920")
+def test_label_whose_b64_representation_is_invalid_utf8():
```
Comment (cygnusv) [inline tests/characters/control/test_web_control.py]: Why is this still skipped?

File: nucypher/characters/lawful.py [hunk 7]
```diff
@@ -296,7 +296,7 @@ def grant(self,
         # value and expiration combinations on a limited number of Ursulas;
         # Users may decide to inject some market strategies here.
         #
-        # TODO: #289
+        # TODO: 289
```
Comment (cygnusv) [inline nucypher/characters/lawful.py]: I think @KPrasch and you have been fighting a silent war: he puts the number sign in TODOs and you remove it.

File: nucypher/network/nodes.py [hunk 8]
```diff
@@ -778,6 +778,7 @@ def learn_from_teacher_node(self, eager=True):
         new_nodes = []
         for node in node_list:
             if not set(self.learning_domains).intersection(set(node.serving_domains)):
+                self.log.debug(f"Teacher {node} is serving {node.serving_domains}, but we're only learning {self.learning_domains}. any of our domains.")
```
Comment (cygnusv) [inline nucypher/network/nodes.py]:
```suggestion
                self.log.debug(f"Teacher {node} is serving {node.serving_domains}, but we're only learning {self.learning_domains}.")
```

File: tests/fixtures.py [hunk 9]
```diff
@@ -284,9 +284,19 @@ def enacted_blockchain_policy(idle_blockchain_policy, blockchain_ursulas):
...
+        def __init__(self):
+            self.messages = []
+            self()
+
+        def __call__(self):
+            enrico = Enrico(policy_encrypting_key=enacted_federated_policy.public_key)
+            message = "Welcome to the flippering.".format(len(self.messages)).encode()
```
Comment (cygnusv) [inline tests/fixtures.py]: why are you doing `.format(len(self.messages))`? It seems to have no effect.

**Comment by cygnusv** [comment]
@jMyles Can you add a TODO in this line: https://github.com/nucypher/nucypher/blob/c613bcd0773cc9ea829b6b948f74755e94bf5769/nucypher/characters/lawful.py#L650, referencing issue #999 ?

File: tests/fixtures.py (hunk 3 above)
Comment (jMyles) [reply tests/fixtures.py]: This is a replacement; we are net zero side channels here.  :-)

File: nucypher/network/server.py (hunk 4 above)
Comment (jMyles) [reply nucypher/network/server.py]: It's not a *specific* issue; it's just that *any* exception that makes it this far needs to be handled differently.

File: nucypher/characters/lawful.py [hunk 10]
```diff
@@ -681,53 +681,71 @@ def join_policy(self, label, alice_verifying_key, node_list=None, block=False):
...
             receiving=self.public_keys(DecryptingPower),
             verifying=alice_verifying_key)
 
-        hrac, map_id = self.construct_hrac_and_map_id(alice_verifying_key, label)
-        _unknown_ursulas, _known_ursulas, m = self.follow_treasure_map(map_id=map_id, block=True)
+        cleartexts = []
 
-        # TODO: Consider blocking until map is done being followed.
+        if must_do_new_retrieval:
+            # TODO: Consider blocking until map is done being followed.
```
[the remark this answers is not in the document]
Comment (jMyles) [reply nucypher/characters/lawful.py]: I guess if it's still of interest - and I think it is - we make an Issue for it.

File: tests/characters/control/test_web_control.py (hunk 6 above)
Comment (jMyles) [reply tests/characters/control/test_web_control.py]: That's a mistake.

File: nucypher/characters/lawful.py (hunk 7 above)
Comment (jMyles) [reply nucypher/characters/lawful.py]: https://getyarn.io/yarn-clip/b855c713-3d55-45c6-8c58-feb088baba55

**Comment by vepkenez** [comment]
I love this PR. I have personally been affected by many of these issues and I'm glad to see them gone! Yay!  💯 👍

File: nucypher/characters/lawful.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (cygnusv) [reply nucypher/characters/lawful.py]:
```suggestion
            # TODO: Consider blocking until map is done being followed. #1114 
```

**Pull request merged**